### PR TITLE
Warn baseclass

### PR DESCRIPTION
**Pull request by ronpal**
# Description

Moving Clean warnings to new Warning classes

## Checklist

- [ ] Tests added/updated.
- [ ] Run Demo Job Locally.
- [ ] Documentation updated.
- [ ] Changelogs updated in [CHANGELOG.cdf-tk.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.cdf-tk.md).
- [ ] Template changelogs updated in [CHANGELOG.templates.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.templates.md).
- [ ] Version bumped.
  [_version.py](https://github.com/cognitedata/toolkit/blob/main/cognite/cognite_toolkit/_version.py) and
  [pyproject.toml](https://github.com/cognitedata/toolkit/blob/main/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite_toolkit/_cdf_tk/warnings.py [hunk 1]
```diff
@@ -6,8 +6,20 @@
 from collections import UserList
 from collections.abc import Collection
 from dataclasses import dataclass
+from enum import Enum
 from functools import total_ordering
-from typing import Any, Generic, TypeVar
+from typing import Any, Generic, TypeVar, Union
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/warnings.py:11]:
```suggestion
from typing import Any, Generic, TypeVar, Union, ClassVar
```

File: cognite_toolkit/_cdf_tk/warnings.py [hunk 2]
```diff
@@ -52,3 +64,40 @@ def __str__(self) -> str:
             for warning in group_list:
                 output.append(f"{'    ' * 2} * {warning!s}")
         return "\n".join(output)
+
+
+@dataclass(frozen=True)
+class GeneralWarning(ToolkitWarning):
+    severity: SeverityLevel | None = None
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/warnings.py:71]:
```suggestion
    severity: ClassVar[SeverityLevel] 
```

File: cognite_toolkit/_cdf_tk/warnings.py [hunk 3]
```diff
@@ -52,3 +64,40 @@ def __str__(self) -> str:
...
+                output.append(f"{'    ' * 2}{self.details}")
+            else:
+                for detail in self.details:
+                    output.append(f"{'    ' * 2}{detail}")
+        return "\n".join(output)
+
+
+@dataclass(frozen=True)
+class ToolkitDependenciesIncludedWarning(GeneralWarning):
+    severity: SeverityLevel = SeverityLevel.LOW
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/warnings.py:89]:
```suggestion
    severity: ClassVar[SeverityLevel] = SeverityLevel.LOW
```

File: cognite_toolkit/_cdf_tk/warnings.py [hunk 4]
```diff
@@ -52,3 +64,40 @@ def __str__(self) -> str:
...
+    message: str = "Some resources were added due to dependencies."
+    details: Union[None, str, list[str]] = None  # Allow None, str, list[str]
+
+    def __init__(self, details: Union[None, str, list[str]] = None) -> None:
+        super().__init__(message=self.message, details=details, severity=self.severity)
+
+
+@dataclass(frozen=True)
+class ToolkitCleanDatasetNotSupportedWarning(GeneralWarning):
+    severity: SeverityLevel = SeverityLevel.LOW
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/warnings.py:99]:
```suggestion
    severity: ClassVar[SeverityLevel] = SeverityLevel.LOW
```

File: cognite_toolkit/_cdf_tk/warnings.py (hunk 2 above)
Comment (doctrino) [reply cognite_toolkit/_cdf_tk/warnings.py:71]: Like we discussed, lets keep severity as a property of the class itself and not the instances.

File: cognite_toolkit/_cdf_tk/warnings.py [hunk 5]
```diff
@@ -52,3 +64,40 @@ def __str__(self) -> str:
             for warning in group_list:
                 output.append(f"{'    ' * 2} * {warning!s}")
         return "\n".join(output)
+
+
+@dataclass(frozen=True)
+class GeneralWarning(ToolkitWarning):
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/warnings.py:70]: Maybe making this an abstract class to signal it is not intended to be used, but instead derived classes.
```suggestion
class GeneralWarning(ToolkitWarning, ABC):
```

File: cognite_toolkit/_cdf_tk/warnings.py [hunk 6]
```diff
@@ -52,3 +64,40 @@ def __str__(self) -> str:
...
+
+    def __str__(self) -> str:
+        output = [f"{RICH_WARNING_FORMAT}{self.severity}{type(self).__name__}: {self.message}"]
+
+        if self.details:
+            if isinstance(self.details, str):
+                output.append(f"{'    ' * 2}{self.details}")
+            else:
+                for detail in self.details:
+                    output.append(f"{'    ' * 2}{detail}")
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/warnings.py:83]: Like it, good flexibility of the developer using of warnings.

File: cognite_toolkit/_cdf_tk/warnings.py [hunk 7]
```diff
@@ -52,3 +64,40 @@ def __str__(self) -> str:
...
+        return "\n".join(output)
+
+
+@dataclass(frozen=True)
+class ToolkitDependenciesIncludedWarning(GeneralWarning):
+    severity: SeverityLevel = SeverityLevel.LOW
+    message: str = "Some resources were added due to dependencies."
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/warnings.py:90]: Not sure if this is possible, but in this case message is a property of the warning class and not intended to be changed when you use the warning.
```suggestion
    message: ClassVar[str] = "Some resources were added due to dependencies."
```

File: cognite_toolkit/_cdf_tk/warnings.py [hunk 8]
```diff
@@ -52,3 +64,40 @@ def __str__(self) -> str:
...
+    details: Union[None, str, list[str]] = None  # Allow None, str, list[str]
+
+    def __init__(self, details: Union[None, str, list[str]] = None) -> None:
+        super().__init__(message=self.message, details=details, severity=self.severity)
+
+
+@dataclass(frozen=True)
+class ToolkitCleanDatasetNotSupportedWarning(GeneralWarning):
+    severity: SeverityLevel = SeverityLevel.LOW
+    message: str = "Dataset cleaning is not supported, skipping..."
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/warnings.py:100]: Same here.

File: cognite_toolkit/_cdf_tk/warnings.py [hunk 9]
```diff
@@ -52,3 +64,40 @@ def __str__(self) -> str:
...
+    severity: SeverityLevel = SeverityLevel.LOW
+    message: str = "Some resources were added due to dependencies."
+    details: Union[None, str, list[str]] = None  # Allow None, str, list[str]
+
+    def __init__(self, details: Union[None, str, list[str]] = None) -> None:
+        super().__init__(message=self.message, details=details, severity=self.severity)
+
+
+@dataclass(frozen=True)
+class ToolkitCleanDatasetNotSupportedWarning(GeneralWarning):
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/warnings.py:98]: Maybe a less specific warning?
```suggestion
class ToolkitNotSupportedWarning(GeneralWarning):
    functionality: str
```

Comment (doctrino) [reply cognite_toolkit/_cdf_tk/warnings.py:98]: Then, instantiate it with `ToolkitNotSupportedWarning("Dataset cleaning")`?

File: cognite_toolkit/_cdf_tk/warnings.py [hunk 10]
```diff
@@ -52,3 +64,40 @@ def __str__(self) -> str:
             for warning in group_list:
                 output.append(f"{'    ' * 2} * {warning!s}")
         return "\n".join(output)
+
+
+@dataclass(frozen=True)
+class GeneralWarning(ToolkitWarning):
+    severity: SeverityLevel | None = None
+    message: str | None = None
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/warnings.py:72]: I think maybe this should be a method instead. That will give more flexibility. 
```python
@dataclass(frozen=True)
class GeneralWarning(ToolkitWarning):
    @abstractmethod
    def get_message() -> str:
         raise NotImplementedError()
```

Then
```python
class ToolkitNotSupportedWarning(GeneralWarning):
    functionality: str

   def get_message(self) -> str:
         return f"{self.functionality} is not supported, skipping...
```

File: cognite_toolkit/_cdf_tk/warnings.py (hunk 7 above)
Comment (doctrino) [reply cognite_toolkit/_cdf_tk/warnings.py:90]: In addition, should probably include which dependencies are included?

**Review by doctrino — COMMENTED** [review]
A few things to discuss.

File: cognite_toolkit/_cdf_tk/warnings.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
+from __future__ import annotations
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/warnings.py:1]: The module should not be called `warnings` as that is the name of a built in module in python

File: cognite_toolkit/_cdf_tk/warnings.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+class SeverityLevel(Enum):
+    HIGH = "HIGH"
+    MEDIUM = "MEDIUM"
+    LOW = "LOW"
+
+
+class SeverityFormat:
+    @staticmethod
+    def get_rich_severity_format(severity: SeverityLevel, *messages: str) -> str:
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/warnings.py:25]: I am a bit skeptical about this one. I think my main issues are the following
* You have to pass in severity.
* Messages are enforced to be given in order. 

But I am not entirely convinced that I am right, so will approve and lets start using it. If it turns out to be error prone we can always change this mechanism.

**Review by doctrino — APPROVED** [review]
One critical `warnings` module needs a rename, but otherwise good to go. Pre-approving.

File: tests/test_output.py [hunk 13]
```diff
@@ -0,0 +1,12 @@
+import subprocess
+
+import pytest
+
+
+@pytest.mark.skip(reason="Not implemented yet")
+def test_print_clean_warnings() -> None:
+    clean_cmd = ["cdf-tk", "deploy", "--env", "dev", "--dry-run"]
+    output = subprocess.run(clean_cmd, capture_output=True, shell=True)
```
Comment (haakonvt) [inline tests/test_output.py:9]: You can use `text=True` to avoid decoding after (I believe)

**Pull request merged**